### PR TITLE
[CodeGen] Speed up ReachingDefAnalysis (NFC)

### PR DESCRIPTION
**Pull request by kazutakahirata**
This patch speeds up ReachingDefAnalysis without changing the
functionality.

Without this patch, ReachingDefAnalysis maintains a three-dimensional
vector to keep track of reaching definitions (i.e. instructions) for
each basic block and each reg unit.  Quite literally:

  MBBReachingDefs[MBBNumber][Unit]

is a vector of definitions, where the innermost vector is implemented
with TinyPtrVector.  The problem is that for many pairs of MBBNumber
and Unit, we have more than one definition, causing TinyPtrVector to
allocate SmallVector, accounting for 1.91% of heap allocations during
the compilation of X86ISelLowering.cpp.ll, a .ll version of
X86ISelLowering.cpp.

This patch replaces the three-dimensional vector with a custom
multimap implemented on a single instance of
std::vector<std::pair<int, int>>.  The vector contains a collection of
singly-linked lists of reaching definitions for all possible pairs of
MBBNumber and Unit.

With this patch, the instruction count and cycle count go down by
0.24% and 0.82%, respectively, for the compilation of
X86ISelLowering.cpp.ll on a stabilized x86 machine (the frequency
govenor set to performance and hyperthreading disabled).  The number
of heap allocations goes down by 2.20%.

### CODE REVIEW

**Comment by s-barannikov** [comment]
How did peak memory consumption change?

**Comment by nikic** [comment]
Right, RDA is more constrained by memory usage than performance. This patch increases max-rss by 2-4% (see "link only" results on https://llvm-compile-time-tracker.com/compare.php?from=4202634886b9b8f1ee74713a8ed8c86941891853&to=afa915730473facf8e9879e00ee0fa5d67e13731&stat=max-rss&linkStats=on).

I have some old patches that I never got around to landing to swap the order of reg unit and BB number storage (basically https://github.com/nikic/llvm-project/commits/perf/reaching-defs-6/) which showed promising results at the time.

**Comment by kazutakahirata** [comment]
> How did peak memory consumption change?

According to heaptrack, the peak memory consumption goes up from 119.4MB to 124.4MB.

According to `/usr/bin/time/ -v`, the max RSS goes up from 286,436KB to 287,972KB.

I could reduce the numbers a little bit by reserving fewer elements for `Storage`.  The current patch sets aside enough storage for `NumBlockIDs * NumRegUnits` reaching definitions.  In 95% cases, we have no more than half of those (that is, `NumBlockIDs * NumRegUnits / 2` reaching definitions).

File: llvm/include/llvm/CodeGen/ReachingDefAnalysis.h [hunk 1]
```diff
@@ -33,35 +32,156 @@ namespace llvm {
...
+
+  // Initialize the multimap with the number of basic blocks and the number of
+  // register units.
+  void init(unsigned BBs, unsigned Regs) {
+    assert(NumBlockIDs == 0 && "can initialize only once");
+    assert(NumRegUnits == 0 && "can initialize only once");
+    assert(Storage.empty() && "can initialize only once");
+    NumBlockIDs = BBs;
+    NumRegUnits = Regs;
+    unsigned NumIndexes = NumBlockIDs * NumRegUnits;
```
Comment (arsenm) [inline llvm/include/llvm/CodeGen/ReachingDefAnalysis.h:72]: For reference, AMDGPU currently has 1409 reg units

File: llvm/include/llvm/CodeGen/ReachingDefAnalysis.h [hunk 2]
```diff
@@ -33,35 +32,156 @@ namespace llvm {
...
+      // Update the index of the first element.
+      Storage[Key].first = NewIndex;
+      // Update the index of the last element.
+      Storage[Key].second = NewIndex;
+    } else {
+      unsigned OldLastPos = Storage[Key].second;
+      // The old last element now points to the new element.
+      Storage[OldLastPos].second = NewIndex;
+      // Update the index of the last element.
+      Storage[Key].second = NewIndex;
```
Comment (arsenm) [inline llvm/include/llvm/CodeGen/ReachingDefAnalysis.h:104]: Doing this map lookup twice? Just do it once

File: llvm/include/llvm/CodeGen/ReachingDefAnalysis.h [hunk 3]
```diff
@@ -33,35 +32,156 @@ namespace llvm {
...
+  // definitions for (MBBNumber, Unit).
+  void prependDef(unsigned MBBNumber, unsigned Unit, int Def) {
+    unsigned Key = computeKey(MBBNumber, Unit);
+    unsigned NewIndex = Storage.size();
+    Storage.emplace_back(Def, 0);
+    if (Storage[Key].first == 0) {
+      // Update the index of the first element.
+      Storage[Key].first = NewIndex;
+      // Update the index of the last element.
+      Storage[Key].second = NewIndex;
```
Comment (arsenm) [inline llvm/include/llvm/CodeGen/ReachingDefAnalysis.h:118]: Looking up twice?

File: llvm/include/llvm/CodeGen/ReachingDefAnalysis.h (hunk 1 above)
Comment (kazutakahirata) [reply llvm/include/llvm/CodeGen/ReachingDefAnalysis.h:72]: Wow!  That's a lot.  Thank you for the data point.  Do you happen to know how many reg units are typically in use in a function?  I see about 25% or less for X86.  (Of course, we might see more utilization if you use 512-bit vectors like AVX-512.)

FWIW, I have a local patch that takes an idea of @nikic's where we only allocate arrays for reg units in use.

Comment (arsenm) [reply llvm/include/llvm/CodeGen/ReachingDefAnalysis.h:72]: I have no idea, but I regularly see programs use nearly all available registers

**Comment by kazutakahirata** [comment]
I'm closing this in favor of developments on top of #110432.

**Pull request closed without merging**